### PR TITLE
Consolidate all uses of the fast range mapping technique in util/fastrange.h

### PR DESCRIPTION
**Pull request by sipa**
Several places in the codebase use the fast range mapping technique described in https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/, some for 32-bit ranges, some for 64-bit ones.

Move all of these to `util/fastrange.h`, and give them a consistent name.

### CODE REVIEW

**Review by Sjors — APPROVED** [review]
ACK b6e8ebc361682ba90cfcdd3af0fd39da85d1ca2a

nit: `funcion` -> `function` in last commit message

File: src/util/fastrange.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#ifndef BITCOIN_UTIL_FASTRANGE_H
+#define BITCOIN_UTIL_FASTRANGE_H
+
+#include <cstdint>
+
+// See: https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/
+
+/** Map a value x that is uniformly distributed in the range [0, 2^64) to a
+ *  value uniformly distributed in [0, n) by returning the upper 64 bits of
+ * x * n. */
```
Comment (shaavan) [inline src/util/fastrange.h:14]: A minor nit. Feel free to ignore.

How about putting the reference link after the explanation?

```suggestion
/** Map a value x that is uniformly distributed in the range [0, 2^64) to a
 *  value uniformly distributed in [0, n) by returning the upper 64 bits of
 * x * n. */
 
 // See: https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/
```

As this convention is followed throughout the codebase and seems more logical.

**Review by shaavan — APPROVED** [review]
ACK b6e8ebc361682ba90cfcdd3af0fd39da85d1ca2a

This PR:

- Renames **MapIntoRange** -> **FastRange64**. Moves it to `util/fastrange`.
- Renames **FastMod** -> **FastRange32**. Moves it to `util/fastrange`.
- Uses **FastRange32** consistently throughout the codebase.

I like the refactoring as this makes the functions more consistent in their naming scheme and where they are defined.
I checked these functions are being appropriately used wherever possible.

I had one suggestion, though.

File: src/util/fastrange.h (hunk 1 above)
Comment (sipa) [reply src/util/fastrange.h]: I've rewritten the comments in the last commit.

**Comment by Sjors** [comment]
ACK efab28b06bfaa50c41337e84136cb58437e7ba00

**Review by shaavan — APPROVED** [review]
reACK efab28b06bfaa50c41337e84136cb58437e7ba00

Changes since my last review:
- Rewritten documentation in `util/fastrange.h`.
- Reordered function definition in `util/fastrange.h`.

The new documentation looks a lot better. It’s able to remain concise while also clearly explaining the following code. Also, now the placement of links is a lot more meaningful.

I prefer the new ordering of functions as it feels more logical, and also it follows lexicographic ordering, so that is a plus. **FastRange32** followed by **FastRange64**.

**Review by maflcko — APPROVED** [review]
review ACK efab28b06bfaa50c41337e84136cb58437e7ba00 🍸

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK efab28b06bfaa50c41337e84136cb58437e7ba00 🍸
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhOvAwAjcxhK6fe5biykR3KX9cq8SXp3ipAgZVHUdSrzB/iWtZBBFWOEIgMKSqB
ohj9G24GyaS4Do6O7AKdclWFzcjOh+3XxEJZvKKYvMF5fiu1Xh3FyNvbArS4MBdl
0WkGfc3IMMAFLowO7YdKt11EJQipd93ujnP/6Zc+SopytTWGz4j6VPDMWiRKAvXi
sxk7nGbCgdotTKUo4jFx6CVFvUBuZNYDKO4pzzpJETrRX+VAkD7Fb7JQSQOoy2mP
HSvVXhMz3DcGEIEkoae2mes5+JsxH7sbD71V1GIsIdpuiTyA4tVyBZp5bqyoaFRA
HqE1j87MkUXLtre8pkRECONriruOOHQLTnfwdlMC5WH4AwPE+CvrhTae8d/iIe4b
xNRtVf+SV36WU72JwejRnL2+0YwmE8TkSudm8ctG5c9pWttOSVUyFoLhdCiXRz+U
Of34WyEpvZO637cqD8no2zbXM3sBjacU+qV0LvYY3p/Kko4qZDDQDUH4A1ip2KI+
QXlBjryM
=2LES
-----END PGP SIGNATURE-----
```


</details>

**Pull request merged**